### PR TITLE
Add weather api

### PR DESCRIPTION
**Pull request by Lev-Stambler**
- Add API for getting weather data
- Convert cloud percentage to power generated by solar panels
- Add data store for daily weather to decrease API usage
- Add tests
- Refactor to async / await to handle async call

### CODE REVIEW

File: biogrid/libs/weather/src/lib/weather.ts [hunk 1]
```diff
@@ -1,10 +1,91 @@
...
+
+interface WeatherHour {
+  cloudCoverage: CloudCoverage;
+  isDay: boolean;
+}
+interface WeatherHourMap {
+  [formattedDate: string]: WeatherHour;
+}
+
+export class WeatherLib {
```
Comment (RolandNaijuka) [inline biogrid/libs/weather/src/lib/weather.ts:16]: How does this put into consideration the dates selected by the user? Can the api provide data about future dates?

File: biogrid/libs/weather/src/lib/weather.ts [hunk 2]
```diff
@@ -1,10 +1,91 @@
...
+  }
+
+  private formatDate(date: Date): string {
+    let d = new Date(date),
+      month = '' + (d.getMonth() + 1),
+      day = '' + d.getDate(),
+      year = d.getFullYear();
+
+    if (month.length < 2) month = '0' + month;
+    if (day.length < 2) day = '0' + day;
```
Comment (RolandNaijuka) [inline biogrid/libs/weather/src/lib/weather.ts]: nit: use Template strings(es6 backticks) to concat the strings

File: biogrid/libs/weather/src/lib/weather.ts [hunk 3]
```diff
@@ -1,10 +1,91 @@
...
+    if (!this.weatherData[dateFormatted]) {
+      throw new Error('Date not found within the specified time range');
+    }
+    return this.weatherData[dateFormatted].cloudCoverage;
+  }
+
+  isDay(date: Date): boolean {
+    const dateFormatted = this.formatDateWithHour(date);
+        if (!this.weatherData[dateFormatted]) {
+      throw new Error('Date not found within the specified time range');
```
Comment (RolandNaijuka) [inline biogrid/libs/weather/src/lib/weather.ts]: could you send the date as part of the Error
such as `` `${date} not found within the specified time range` ``
also clarify on the time range

File: biogrid/libs/weather/src/lib/weather.ts [hunk 4]
```diff
@@ -1,10 +1,91 @@
...
+  isSetup() {
+    return this.setupOccured;
+  }
+
+  getCloudCoverage(date: Date): CloudCoverage {
+    const dateFormatted = this.formatDateWithHour(date);
+    if (!this.weatherData[dateFormatted]) {
+      throw new Error('Date not found within the specified time range');
+    }
+    return this.weatherData[dateFormatted].cloudCoverage;
```
Comment (RolandNaijuka) [inline biogrid/libs/weather/src/lib/weather.ts]: could you send the date as part of the Error
such as `` `${date} not found within the specified time range` ``
also clarify on the time range

File: biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts [hunk 5]
```diff
@@ -63,16 +74,16 @@ export class SolarPanel extends EnergySource {
...
+  async isEmpty() {
+    return (await this.getEnergyInJoules()) === 0;
   }
 
-  private intensityToKiloWattsPerSquareMeter(intensity: SunlightIntensity) {
-    // Calculation derived from https://www.researchgate.net/post/Howto_convert_solar_intensity_in_LUX_to_watt_per_meter_square_for_sunlight
-    return SOLAR_PANEL.KILOLUX_TO_KILOWATT_PER_SQUARE_METER * intensity;
+  private cloudCoverageToKiloWattsPerSquareMeter(cloudCoverage: CloudCoverage) {
+    // CalculationDerived from https://scool.larc.nasa.gov/lesson_plans/CloudCoverSolarRadiation.pdf
+    return 990 * (1 - 0.75 * Math.pow(cloudCoverage, 3)) / 1000;
```
Comment (RolandNaijuka) [inline biogrid/libs/biogrid-simulator/src/lib/bioenergy-source/solar-panel.ts]: nit: include these constants in the `bio-constants` file

**Review by RolandNaijuka — CHANGES_REQUESTED** [review]

File: biogrid/libs/weather/src/lib/weather.ts (hunk 1 above)
Comment (Lev-Stambler) [reply biogrid/libs/weather/src/lib/weather.ts:16]: The weather lib is just for the grid to use with whatever date it needs to specify

File: biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts [hunk 6]
```diff
@@ -114,37 +99,36 @@ export class BioBrain implements Brain {
...
+    let solarPanels: SolarPanel[] = gridItems[GRID_ITEM_NAMES.SOLAR_PANEL];
     // Assuming the large battery is not fully charged
     largeBatteries = largeBatteries.filter((battery) => !battery.isFull());
 
     // Filter the solar panels and remove the ones with the minimum energy or empty
     const solarPanelsFiltered = await this.filterSolarPanelsByEnergyAmount(solarPanels)
 
     // Create an array of the possible energy givers
-    const allEnergyProviders = [...solarPanelsFiltered];
+    const allEnergyProviders = [...solarPanels];
```
Comment (RolandNaijuka) [inline biogrid/libs/biogrid-simulator/src/lib/biobrain/biobrain-class.ts:114]: I see you changed the filter solar panels but did not collect this part. can you insert the right variable which is the filtered solar panels

**Review by RolandNaijuka — CHANGES_REQUESTED** [review]

File: biogrid/libs/weather/src/lib/weather.ts (hunk 1 above)
Comment (RolandNaijuka) [reply biogrid/libs/weather/src/lib/weather.ts:16]: alright

**Review by RolandNaijuka — APPROVED** [review]

**Pull request merged**